### PR TITLE
test for host cleanup in unfavorable pod deletes

### PR DESCRIPTION
**Pull request by jeffvance**
addresses issue #31272 with a new e2e test in _kubelet.go_
```release-note
NONE
```

### CODE REVIEW

**Comment by jeffvance** [comment]
@saad-ali @jsafrane PTAL when you get a chance. Thanks!

**Comment by jeffvance** [comment]
adding @pmorie

File: test/e2e/kubelet.go [hunk 1]
```diff
@@ -127,8 +131,142 @@ func updateNodeLabels(c clientset.Interface, nodeNames sets.String, toAdd, toRem
...
+	check := func(e int) {
+		if expectClean {
+			Expect(e).NotTo(BeZero())
+		} else {
+			Expect(e).To(BeZero())
+		}
+	}
+
+	nodeName := pod.Spec.NodeName
+	if len(nodeName) == 0 {
```
Comment (jeffvance) [inline test/e2e/kubelet.go]: When deployed locally both _pod.Spec.NodeName_ and _pod.Status.PodUID_ are empty. Is this expected? Is there another way to get the node name of the target pod's host? 
Note: `framework.SSH()` seems to work when the node name is empty.

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -3630,15 +3631,29 @@ func GetSigner(provider string) (ssh.Signer, error) {
...
 		keyfile = "kube_aws_rsa"
 	case "vagrant":
-		keyfile := os.Getenv("VAGRANT_SSH_KEY")
+		keyfile = os.Getenv("VAGRANT_SSH_KEY")
 		if len(keyfile) != 0 {
 			return sshutil.MakePrivateKeySignerFromFile(keyfile)
 		}
 		return nil, fmt.Errorf("VAGRANT_SSH_KEY env variable should be provided")
+	case "local":
+		keyfile = os.Getenv("LOCAL_SSH_KEY") // maybe?
```
Comment (jeffvance) [inline test/e2e/framework/util.go]: do we want to support a LOCAL_SSH_KEY var for local?

**Comment by jeffvance** [comment]
@saad-ali, @jsafrane  I am getting a _permission denied_ error when trying to execute `ls` command over SSH. Here is the logfile fragment:
```
I1214 10:28:32.557] STEP: Check if host running deleted pod has been cleaned up -- expect not
I1214 10:28:32.557] Dec 14 10:28:20.550: INFO: ssh jenkins@35.184.85.28:22: command:   ls /var/lib/kubelet/pods/027cd4ff-c22b-11e6-a7e5-42010af00002
I1214 10:28:32.557] Dec 14 10:28:20.550: INFO: ssh jenkins@35.184.85.28:22: stdout:    ""
I1214 10:28:32.558] Dec 14 10:28:20.550: INFO: ssh jenkins@35.184.85.28:22: stderr:    "ls: cannot access /var/lib/kubelet/pods/027cd4ff-c22b-11e6-a7e5-42010af00002: Permission denied\n"
```
Do you have any suggestion on how to fix this? BTW, the same `ls` command over SSH works in my own e2e testing using GCE deployment.  Thanks!

**Comment by jeffvance** [comment]
Test failure analysis:
1) https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/38214/pull-kubernetes-e2e-gce/9961/?log#log --
After the nfs server pod is started again, the test expects the nfs mount and pod's uid dir to be deleted from the host running the pod. This is done by ssh-ing to the host and executing `sudo ls /var/lib/kubelet/pods/68991429-c241-11e6-b343-42010af00002` and expecting an exitCode != 0, meaning there are no sub-dirs under that dir. Instead, we see in the log that 4 sub-dirs are listed, and thus the test fails.
2) https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/38214/pull-kubernetes-e2e-gce-etcd3/9464/?log#log -- ditto from above.
3) https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/38214/pull-kubernetes-e2e-gce-gci/9437/ -- the wait period for the pod (binding to the nfs mount) to start is exceeded. This is likely a test env issue since the same test works in my gce deployment.

**Question:** for tests 1  and 2, should there be an explicit delay before testing that the host has been cleaned up? The only delay now is the time it takes for the nfs-server pod to start up.

**Comment by jeffvance** [comment]
@saad-ali @jsafrane @childsb It appears to me that this test is working as expected and that the failure is a result of the pod's host not being cleaned up. By "cleaned up" I am referring to the nfs mount and the pod's uid dir are still present on the host. Please advise on the next step so that this test PR can be merged.

File: test/e2e/kubelet.go [hunk 3]
```diff
@@ -228,4 +379,91 @@ var _ = framework.KubeDescribe("kubelet", func() {
...
+	// Delete nfs server pod after another pods accesses the mounted nfs volume.
+	framework.KubeDescribe("host cleanup with volume mounts [HostCleanup]", func() {
+		var (
+			nfsServerPod *v1.Pod
+			nfsIP        string
+		)
+		NFSconfig := VolumeTestConfig{
+			namespace:   v1.NamespaceDefault,
+			prefix:      "nfs",
+			serverImage: "gcr.io/google_containers/volume-nfs:0.7",
```
Comment (copejon) [inline test/e2e/kubelet.go]: latest version is 0.8

File: test/e2e/kubelet.go [hunk 4]
```diff
@@ -128,59 +131,207 @@ func updateNodeLabels(c clientset.Interface, nodeNames sets.String, toAdd, toRem
...
+		},
+		ObjectMeta: v1.ObjectMeta{
+			GenerateName: "pod-nfs-vol-",
+			Namespace:    ns,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    "pod-nfs-vol",
+					Image:   "gcr.io/google_containers/busybox:1.24",
```
Comment (jeffvance) [inline test/e2e/kubelet.go]: Agree. I think maybe a new PR should be created with the sole purpose of creating these constants for the nfs-server pod, busybox, and storage containers such as glusterfs and ceph.

**Comment by jsafrane** [comment]
> Question: for tests 1 and 2, should there be an explicit delay before testing that the host has been cleaned up? The only delay now is the time it takes for the nfs-server pod to start up.

@jeffvance, when you delete a pod from API server, kubelet *starts* cleaning it up. It may take some time until it unmounts everything and removes all directories. You must check presence of /var/lib/kubelet/pods/<uid> several times in a loop, just like e.g. `WaitForPodReady`. 5 minutes of total wait could be a good start, kubelet is not particularly fast and test machines are heavily overcommitted.

**Comment by jeffvance** [comment]
working on other e2e pv test issues. Will rebase and implement Jan's suggestion as soon as I free up...

**Comment by jeffvance** [comment]
@jsafrane I've implemented a wait loop for testing if the pod's uid dir and its nfs mount have been removed from the host minion. Even so, after 5m wait the host has still not unmounted the nfs mount nor removed the uid dir. I waited manually more than 20 min with same result:
```
root@jeffvance-e2e-minion-group-7szh:/home/jvance# mount|grep nfs
rpc_pipefs on /var/lib/nfs/rpc_pipefs type rpc_pipefs (rw,relatime)
10.244.2.8:/exports on /var/lib/kubelet/pods/6d9a0dcc-dc36-11e6-98b6-42010af00002/volumes/kubernetes.io~nfs/nfs-vol
 type nfs (rw,relatime,vers=3,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,mou
ntaddr=10.244.2.8,mountvers=3,mountport=20048,mountproto=udp,local_lock=none,addr=10.244.2.8)
```
So, I believe there may still be a minion cleanup issue for mounts, or perhaps some irregularities with the GCE environment?  Thoughts on this? Thanks

**Comment by jeffvance** [comment]
@jsafrane @saad-ali Looks like CI passes now so PTAL. This pr is an important part of verifying a "wedged" kubelet. Thanks. Also, @eparis why is my CLA not being reported? I thought I've completed all of the linux foundation stuff a while ago...

**Comment by jeffvance** [comment]
@copejon PTAL when you get a chance. Thx!

**Comment by jeffvance** [comment]
@saad-ali per sig-storage call this morning, can you ptal? Thanks!

**Comment by saad-ali** [comment]
@jingxu97 Could you please review this PR

File: test/e2e/kubelet.go [hunk 5]
```diff
@@ -228,4 +404,87 @@ var _ = framework.KubeDescribe("kubelet", func() {
 			})
 		}
 	})
+
+	// Delete nfs server pod after another pods accesses the mounted nfs volume.
+	framework.KubeDescribe("host cleanup with volume mounts [HostCleanup]", func() {
```
Comment (gnufied) [inline test/e2e/kubelet.go]: Since this code specifically tests a particular volume plugin, how hard it is to move this into its own file? such `kubelet_nfs.go` or something?

**Review by gnufied — COMMENTED** [review]
We should also add following tags IMO:

1. `[volume]` tag
2.  `[Slow]` tag

**Review by jeffvance — COMMENTED** [review]
x

File: test/e2e/kubelet.go [hunk 6]
```diff
@@ -92,6 +97,22 @@ func waitTillNPodsRunningOnNodes(c clientset.Interface, nodeNames sets.String, p
...
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(poll) {
+		done, err := cond()
+		if done {
+			return err
+		}
+		framework.Logf("Waiting another %v for host to be in expected condition", poll)
+	}
+
+	return fmt.Errorf("gave up after waiting %v for kubelet's host to be in expected condition", timeout)
+}
```
Comment (jeffvance) [inline test/e2e/kubelet.go]: @gnufied I added [Slow] to kubelet.go -- good suggestion and I meant to do that but forgot. I think, as of now, all of the pv related e2e tests are slow, some with the possibility of 5m waits. I have not added [Slow] to the test in volumes.go since I did not write them and have not monitored their speed.

I will use an existing pattern for the wait loop.

File: test/e2e/kubelet.go (hunk 5 above)
Comment (jeffvance) [inline test/e2e/kubelet.go]: @gnufied It may turn out to be a good idea to create a new files for these tests; however, atm I am ok with keeping them in _kubelet.go_. Reasons: 1) using nfs is somewhat incidental since the main purpose of the test is to verify that the kubelet cleans up volume mounts and the pod's uid dir. I don't foresee needing this type of test for every in-tree storage plugin. 2) some of the existing tests in _kubelet.go_ are disruptive and so it seems that mine fits ok there, at least for now.

**Comment by jeffvance** [comment]
@jsafrane @copejon @gnufied ptal. I've addressed the comments. Thanks. cc @saad-ali

**Comment by gnufied** [comment]
lgtm

**Comment by jeffvance** [comment]
@jsafrane @gnufied @copejon PTAL, would really like to get this merged. Other work is based on this. Thanks.  cc @saad-ali

File: test/e2e/framework/util.go [hunk 7]
```diff
@@ -376,10 +376,10 @@ func SkipIfMissingResource(clientPool dynamic.ClientPool, gvr schema.GroupVersio
 }
 
 // ProvidersWithSSH are those providers where each node is accessible with SSH
-var ProvidersWithSSH = []string{"gce", "gke", "aws"}
+var ProvidersWithSSH = []string{"gce", "gke", "aws", "local"}
```
Comment (jsafrane) [inline test/e2e/framework/util.go]: adding ssh support  to local tests seems unrelated to this PR, please post a separate one.

File: test/e2e/volumes.go [hunk 8]
```diff
@@ -161,17 +162,35 @@ func startVolumeServer(client clientset.Interface, config VolumeTestConfig) *v1.
...
-	framework.ExpectNoError(err, "Cannot locate the server pod %v: %v", serverPod.Name, err)
+	if pod == nil {
+		By(fmt.Sprintf("locating the %q server pod", serverPodName))
+		pod, err = podClient.Get(serverPodName, metav1.GetOptions{})
+		framework.ExpectNoError(err, "Cannot locate the server pod %q: %v", serverPodName, err)
+		//The server pod doesn't do a mount and with the WaitForPod... above the server pod is
+		//running, therefore the sleep below should not be necessary.
+		//TODO: if a delay is actually required it is better to create a test pod which
+		//      mounts the volume and then wait for that pod to terminate.
+		//By("sleeping a bit to give the server time to start")
```
Comment (jsafrane) [inline test/e2e/volumes.go]: A running pod does not indicate that a daemon is actually running inside it. For NFS, the slowdown is negligible, other test images do non-trivial initialization on startup (e.g. gluster and ceph RBD).  I think that kubelet should retry mounting a volume when the server is not available for the first time, however can you post a separate PR with this so it's easier to revert when something goes wrong?

**Review by jsafrane — COMMENTED** [review]
Overall the code looks solid, however I would like to split it into several independent PRs.

**Comment by jeffvance** [comment]
@jsafrane I've split this pr into a total of 3 prs. 1) #40320: _add resillency to the new volume-server func_, and 2) #40319: _ssh support for local_. Now this pr changes just _kubelet.go_ to add tests to see if the kubelet cleans up the host's volume mount and pod's uid dir.  PTAL again. Thanks.  cc. @gnufied @copejon

**Comment by jsafrane** [comment]
@k8s-bot cvm gce e2e test this

**Comment by jeffvance** [comment]
@k8s-bot gci gke e2e test this

**Comment by jsafrane** [comment]
Great, now we're waiting for approval.
@liggitt, @deads2k can you please approve this PR? You seem to be omniponent in test/ directory.

**Comment by deads2k** [comment]
> Great, now we're waiting for approval.
> @liggitt, @deads2k can you please approve this PR? You seem to be omniponent in test/ directory.

@jsafrane Can you experiment with moving volume tests to a subdirectory under e2e so that we set up approvers more suited to particular test buckets?  You just happen to be the first person to ask me, but I think it's a thing that will help.

This may seem a bit silly, but we've had issues with people trying relax conformance tests and the like, so approving tests does make sense in some contexts.

**Comment by jeffvance** [comment]
David, Jan: Jon Cope and I have discussed moving the persistent volumes tests into their own dir. 
We are interested in this idea will create a new pr for this and some smaller, related restructuring.


----- Original Message -----

**Comment by jeffvance** [comment]
I am not able to add a label to this pr. Can @jsafrane or @deads2k please add the label "release-note-none". Thanks.

**Comment by jeffvance** [comment]
@deads2k Thank you for dealing with the labels. I see it is still not being merged. Is there a trigger I can pull to move this along? Thanks.

**Comment by deads2k** [comment]
@eparis any idea why the bot removed my approval?

**Comment by grodrigues3** [comment]
@deads2k not sure why just yet.  I'm looking into it

**Comment by grodrigues3** [comment]
@deads2k Found the bug.  Thanks for your patience David
